tidy3d/components/medium.py

```diff
@@ -1267,6 +1336,257 @@ def elements(self) -> Dict[str, IsotropicMediumType]:
+        rotation : RotationType
+                Rotation applied to diagonal permittivity tensor.
+
+        Returns
+        -------
+        :class:`FullyAnisotropicMedium`
+            Resulting fully anisotropic medium.
+        """
+
+        permittivity_diag = [
+            [xx.permittivity, 0, 0],
+            [0, yy.permittivity, 0],
+            [0, 0, zz.permittivity],
+        ]
+
+        conductivity_diag = [
+            [xx.conductivity, 0, 0],
+            [0, yy.conductivity, 0],
+            [0, 0, zz.conductivity],
+        ]
+
+        permittivity = rotation.rotate_tensor(permittivity_diag)
+        conductivity = rotation.rotate_tensor(conductivity_diag)
+
+        return cls(permittivity=permittivity, conductivity=conductivity)
+
+    @cached_property
+    def eps_sigma_diag(self) -> Tuple[Tuple[float, float, float], Tuple[float, float, float]]:
+        """Main components of permittivity and conductivity tensors."""
+
+        perm_diag, vecs = np.linalg.eig(self.permittivity)
+        cond_diag = np.diag(np.matmul(np.transpose(vecs), np.matmul(self.conductivity, vecs)))
+
+        return (perm_diag, cond_diag)
+
+    @ensure_freq_in_range
+    def eps_model(self, frequency: float) -> complex:
+        """Complex-valued permittivity as a function of frequency."""
+        perm_diag, cond_diag = self.eps_sigma_diag
+
+        eps_diag = AbstractMedium.eps_sigma_to_eps_complex(perm_diag, cond_diag, frequency)
+        return np.mean(eps_diag)
+
+    @ensure_freq_in_range
+    def eps_diagonal(self, frequency: float) -> Tuple[complex, complex, complex]:
+        """Main diagonal of the complex-valued permittivity tensor as a function of frequency."""
+
+        perm_diag, cond_diag = self.eps_sigma_diag
+
+        if np.isscalar(frequency):
+            eps_diag = AbstractMedium.eps_sigma_to_eps_complex(perm_diag, cond_diag, frequency)
+        else:
+            eps_diag = AbstractMedium.eps_sigma_to_eps_complex(
+                perm_diag[:, None], cond_diag[:, None], frequency
+            )
+
+        return eps_diag
+
+    def eps_comp(self, row: Axis, col: Axis, frequency: float) -> complex:
+        """Single component the complex-valued permittivity tensor as a function of frequency."""
+
+        return AbstractMedium.eps_sigma_to_eps_complex(
+            self.permittivity[row][col], self.conductivity[row][col], frequency
+        )
+
+    @cached_property
+    def n_cfl(self):
+        """This property computes the index of refraction related to CFL condition, so that
+        the FDTD with this medium is stable when the time step size that doesn't take
+        material factor into account is multiplied by ``n_cfl``.
+
+        For this medium, it take the minimal of ``sqrt(permittivity)`` for main directions.
+        """
+
+        perm_diag, _ = self.eps_sigma_diag
+        return min(np.sqrt(perm_diag))
+
+    @add_ax_if_none
+    def plot(self, freqs: float, ax: Ax = None) -> Ax:
```

Comment:
this looks almost exactly identical to the `AnistropicMedium.plot()` is there any way to reuse the code there? for example, either:
1. have `AnistropicMedium.plot()` construct a `FullyAnisotropicMedium.from_diagonals()` and call this plotting method.
2. have `FullyAnistropicMedium.plot()` construct an `AnisotropicMedium` and call it's plotting method.

What do you think?

Reply:
Yeah, that's correct, the only difference is that in `AnisotropicMedium` we label components as `("xx", "yy", "zz")`, but for `FullyAnisotropicMedium` I used `("1", "2", "3")`, since main components are oriented arbitratry. I might do what you suggested as (2) and add a check inside the plotting function whether to use `("xx", "yy", "zz")` or `("1", "2", "3")` based on off-diagonal components

Reply:
actually, on the contrary, had to call `AnisotropicMedium.plot()` from inside of `FullyAnisotropicMedium.plot()` because `AnisotropicMedium` is more general in the sense that it allows dispersive materials, while `FullyAnisotropicMedium` currently doesn't